docs/source/design/user_messaging.rst

```diff
@@ -0,0 +1,124 @@
+and are used to inform generic error handling, as well as particular calling commands
+on how to proceed with a specific operation.
+
+Yield result records when:
+
+1. A generic and standard way of error handling is considered useful
+2. A process can reasonably be continued despite evident errors or shortcomings in
+   its subprocesses
+
+Based on the ``status`` field of a result record, a result is categorized into
+*success* (``ok``, ``notneeded``) and *failure* (``impossible``, ``error``).
+Depending on the ``on_failure`` parameterization of a command call (``on_failure='stop'``,
+``on_failure='continue'``, or ``on_failure='ignore'``), any failure-result
+emitted by a command can lead to an ``IncompleteResultsError`` being raised on command
+exit, or a non-zero exit code on the command line.
+
+
+Exception handling
+------------------
+
+In general, **exceptions should be raised when there is no way to ignore or recover from
+the offending action**.
+
+More specifically, raise an exception when:
+
+1. A command's parameter specifications are violated
+2. An additional requirement (beyond parameters) for the successful running of a
+   command, function, or process is not met
+
+It must be made clear to the user/caller what the exact cause of the exception
+is, given the context within which the user/caller triggered the action.
+This is achieved directly via a (re)raised exception, as opposed to logging messages or
+results records which could be ignored or unseen by the user.
+
+If the relevant caller receives a set of result records, these should be inspected
+individually in order to identify result status values that could require an exception
+to be raised to the user/caller. Depending on developer-defined logic, an exception can
+then be raised with an unambiguous message that excludes internal and intermediate
+result messages.
+
+Additionally, in the case of a complex set of dependent actions it could be expensive to
+confirm parameter violations. In such cases, initial sub-processes might already generate
+result records that have to be inspected by the caller, and it could be practically better
+to yield a result record (with ``status=[error|impossible]``) to communicate the failure.
+It would then be up to the upstream caller to decide whether to specify
+``on_failure='ignore'`` or whether to inspect individual result records and turn them
+into exceptions or not.
+
+
+Logging
+-------
+
+Logging provides developers with additional means to describe steps in a process,
+so as to **allow insight into the program flow during debugging** or analysis of e.g.
+usage patterns. Logging can be turned off externally, filtered, and redirected. Apart from
+the log-level and message, it is not inspectable and cannot be used to control the logic
+or flow of a program.
+
+Importantly, logging is not a user messaging method. Therefore:
+
+1. No command should rely on logging for user communication.
+
+
+UI Module
+---------
+
```

Comment:
one earlier comment by @mih would probably fit in well to this paragraph (it came in the context of "we should never write to ``sys.stout`` explicitly):

> We have to use ui.message -- otherwise this message might never see the light of day, because there could be no terminal. ui.message is used to send a message to whatever the UI actually is. Switching to result renderers solves this automatically -- they do that


so maybe relevant points:

- result renderer rely internally on the UI module
- Communication should always happen through the UI module as it provides the flexibility to adjust to the present UI

Reply:
Is there any useful documentation of the UI module, other than found in docstrings and comments, that we can link to?

Reply:
I don't think there is.

Reply:
I think, given that the document says it's a partial description of what is implemented, this part needs to make clear that it is more about what to aim for than what is.
I very much support the aim "No command should rely solely on logging for user communication", but there are a lot violations of this. (I have a "migration" strategy in mind, but that's for another place)


What I would like to add to that section in some form is, that user messaging and logging generally have a different target audience. That's sort of covered by 3., but 1. somewhat blurs the line. Why that's important is: wording of the respective messages may need to be quite different. User-facing messages only devs can understand are training users to ignore output entirely. Which ties into another aspect of my first remark: I'd like to switch to have no logging being spit out by default (write to file instead). What is intended for "regular user" messaging should become a dedicated message via `datalad.ui`. What is not should go into the logger.

Reply:
@bpoldrack your comments are made under the "UI module" section, but seem to refer to logging (referring to the numbers 1, 2 and 3). Can you clarify where exactly you think changes should be made?

Reply:
Indeed - was meant to refer to the 3 points above this section.

Changes:
If we agree on the notion, that this document is actually more about describing the goal than the status quo, then I would:
- strike the "solely" in point 1 under logging
- change "scope and status" of the entire document accordingly
- change 2nd paragraf in UI module:
   "Therefore: ~in cases where existing user communication processes are not appropriate,~
     developers should let explicit user communication happen through the UI module ..."
   Maybe adding "unless the criteria for logging apply"

Crucial part is "If we agree on the notion ...", though.

Reply:
Thanks. IMO most if not all of the document describes the implementation goal, so I will update the scope accordingly and make the 

I don't think the "solely" should be removed, because that helps convey the point: logging can be used as a user messaging method (in addition to debugging), but not as the _sole_ method for a _command_.